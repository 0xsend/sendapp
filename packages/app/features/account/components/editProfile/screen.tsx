import {
  Button,
  Checkbox,
  FadeCard,
  Paragraph,
  ProfileAvatar,
  Separator,
  Spinner,
  SubmitButton,
  XStack,
  YStack,
} from '@my/ui'
import { SchemaForm } from 'app/utils/SchemaForm'
import { ProfileSchema, useProfileMutation } from 'app/utils/useProfileMutation'
import { useUser } from 'app/utils/useUser'
import { UploadAvatar, type UploadAvatarRefObject } from '../uploadProfileImage/screen'
import { useRef, useState } from 'react'
import type { Tables } from '@my/supabase/database.types'
import { Check } from '@tamagui/lucide-icons'
import { SettingsHeader } from 'app/features/account/settings/components/SettingsHeader'
import { ReadOnlyFieldWithLabel } from 'app/features/account/settings/components/ReadOnlyFieldWithLabel'
import { FieldWithLabel } from 'app/features/account/settings/components/FieldWithLabel'
import { useForm } from 'react-hook-form'
import type { z } from 'zod'

enum FormState {
  Overview = 'Overview',
  ProfileForm = 'ProfileForm',
}

export const EditProfile = () => {
  const [formState, setFormState] = useState<FormState>(FormState.Overview)
  const { profile } = useUser()

  return (
    <YStack w={'100%'}>
      <YStack gap={'$3.5'}>
        <SettingsHeader>Profile</SettingsHeader>
        {(() => {
          switch (true) {
            case !profile:
              return <Spinner size="large" />
            case formState === FormState.Overview:
              return (
                <Overview profile={profile} onPress={() => setFormState(FormState.ProfileForm)} />
              )
            case formState === FormState.ProfileForm:
              return (
                <EditProfileForm
                  profile={profile}
                  onSave={() => setFormState(FormState.Overview)}
                />
              )
          }
        })()}
      </YStack>
    </YStack>
  )
}

const Overview = ({ profile, onPress }: { profile: Tables<'profiles'>; onPress: () => void }) => {
  const { name, about, is_public, avatar_url } = profile
  const avatarRef = useRef<UploadAvatarRefObject>(null)

  return (
    <YStack gap={'$5'}>
      <FadeCard>
        <XStack gap={'$5'} width={'100%'}>
          <UploadAvatar ref={avatarRef}>
            <ProfileAvatar
              avatarUrl={avatar_url ? avatar_url : undefined}
              $gtMd={{ size: 88 }}
              size={88}
            />
          </UploadAvatar>
          <YStack gap={'$2'} ai={'flex-start'}>
            <YStack>
              <Paragraph
                size={'$5'}
                fontWeight={500}
                color={'$lightGrayTextField'}
                $theme-light={{ color: '$darkGrayTextField' }}
              >
                Profile Picture
              </Paragraph>
            </YStack>
            <Button unstyled onPress={() => avatarRef.current?.pickImage()}>
              <Button.Text
                textDecorationLine="underline"
                color="$primary"
                $theme-light={{ color: '$color12' }}
                size={'$5'}
              >
                Update
              </Button.Text>
            </Button>
          </YStack>
        </XStack>
        <FieldWithLabel label={'Name'}>
          <Paragraph size={'$8'} fontWeight={'500'}>
            {name || '-'}
          </Paragraph>
        </FieldWithLabel>
        <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
        <ReadOnlyFieldWithLabel label={'About'} text={about || '-'} />
        <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
        <XStack gap={'$3'} ai={'center'}>
          <Checkbox
            disabled={true}
            checked={!!is_public}
            borderWidth={0}
            backgroundColor={is_public ? '$primary' : '$background'}
            circular={true}
          >
            <Checkbox.Indicator>
              <Check color={'$black'} />
            </Checkbox.Indicator>
          </Checkbox>
          <Paragraph size={'$5'}>{`${is_public ? 'Public' : 'Private'}`} Profile</Paragraph>
        </XStack>
      </FadeCard>
      <SubmitButton theme="green" borderRadius={'$4'} p={'$4'} onPress={onPress}>
        <Button.Text ff={'$mono'} fontWeight={'500'} tt="uppercase" size={'$5'} color={'$black'}>
          edit profile
        </Button.Text>
      </SubmitButton>
    </YStack>
  )
}

function EditProfileForm({ profile, onSave }: { profile: Tables<'profiles'>; onSave: () => void }) {
  const { id, name, about, is_public } = profile
  const { mutateAsync, error } = useProfileMutation(id)
  const form = useForm<z.infer<typeof ProfileSchema>>()

  const handleSubmit = async () => {
    const values = form.getValues()
    await mutateAsync(values)
    onSave()
  }

  return (
    <SchemaForm
      form={form}
      schema={ProfileSchema}
      props={{
        name: {
          'aria-label': 'Name',
          bc: '$color0',
        },
        about: {
          'aria-label': 'Bio',
          placeholder: 'Tell us about yourself',
          backgroundColor: '$color0',
          rows: 2,
          focusStyle: {
            fontStyle: 'normal',
          },
        },
        isPublic: {
          defaultChecked: is_public !== null ? is_public : true,
        },
      }}
      defaultValues={{
        name: name ? name : '',
        about: about ? about : '',
        isPublic: is_public !== null ? is_public : true,
      }}
      formProps={{
        $gtSm: {
          maxWidth: '100%',
        },
      }}
      onSubmit={handleSubmit}
      renderAfter={({ submit }) => (
        <YStack>
          <SubmitButton
            theme="green"
            borderRadius={'$4'}
            p={'$4'}
            mt={'$1'}
            onPress={() => submit()}
          >
            <Button.Text
              ff={'$mono'}
              fontWeight={'500'}
              tt="uppercase"
              size={'$5'}
              color={'$black'}
            >
              SAVE CHANGES
            </Button.Text>
          </SubmitButton>
          {error && (
            <Paragraph marginTop={'$3'} theme="red" color="$color9">
              {error.message}
            </Paragraph>
          )}
        </YStack>
      )}
    >
      {({ name, about, isPublic }) => (
        <FadeCard>
          <FieldWithLabel label={'Name'} gap={'$2'}>
            {name}
          </FieldWithLabel>
          <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
          <FieldWithLabel label={'About'} gap={'$2'}>
            {about}
          </FieldWithLabel>
          <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
          <XStack gap={'$3'} ai={'center'}>
            {isPublic}
            <Paragraph size={'$5'}>Make Profile Public</Paragraph>
          </XStack>
        </FadeCard>
      )}
    </SchemaForm>
  )
}
