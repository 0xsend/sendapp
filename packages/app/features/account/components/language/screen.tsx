import { Card, Paragraph, Separator, Spinner, XStack, YStack } from '@my/ui'
import { SettingsHeader } from 'app/features/account/components/SettingsHeader'
import { useTranslation } from 'react-i18next'
import { useCallback, useMemo, useState } from 'react'
import { Pressable, Platform } from 'react-native'
import { Check } from '@tamagui/lucide-icons'
import { getSupportedLocaleOptions } from 'app/i18n/locales'
import { useHoverStyles } from 'app/utils/useHoverStyles'

export const LanguagePreferences = () => {
  const { i18n, t } = useTranslation('settings')
  const [pendingLocale, setPendingLocale] = useState<string | null>(null)
  const resolvedLanguage = i18n.resolvedLanguage ?? i18n.language

  const options = useMemo(() => getSupportedLocaleOptions(), [])
  const hoverStyles = useHoverStyles()

  const handleSelect = useCallback(
    async (locale: string) => {
      if (locale === resolvedLanguage || pendingLocale === locale) {
        return
      }

      setPendingLocale(locale)
      try {
        await i18n.changeLanguage(locale)
      } finally {
        setPendingLocale(null)
      }
    },
    [i18n, pendingLocale, resolvedLanguage]
  )

  return (
    <YStack gap={'$3.5'} w={'100%'}>
      {Platform.OS === 'web' && <SettingsHeader>{t('language.title')}</SettingsHeader>}
      <Card gap={'$3'} padded size={'$4'}>
        <Paragraph size={'$4'} color={'$color12'}>
          {t('language.description')}
        </Paragraph>
        {pendingLocale ? (
          <Paragraph size={'$2'} color={'$lightGrayTextField'}>
            {t('language.updating')}
          </Paragraph>
        ) : null}
        <Separator boc={'$darkGrayTextField'} opacity={0.2} />
        <YStack>
          {options.map((option, index) => {
            const isActive = option.value === resolvedLanguage
            const isPending = pendingLocale === option.value

            return (
              <YStack key={option.value}>
                <Pressable
                  accessibilityRole="button"
                  accessibilityState={{ selected: isActive, busy: isPending }}
                  disabled={isPending}
                  onPress={() => handleSelect(option.value)}
                >
                  <XStack
                    ai="center"
                    jc="space-between"
                    px={'$2'}
                    py={'$3'}
                    br={'$4'}
                    hoverStyle={hoverStyles}
                  >
                    <Paragraph size={'$5'}>{option.autonym}</Paragraph>
                    {isPending ? (
                      <Spinner size="small" />
                    ) : isActive ? (
                      <Check size={16} color={'$primary'} />
                    ) : null}
                  </XStack>
                </Pressable>
                {index < options.length - 1 && (
                  <Separator boc={'$darkGrayTextField'} opacity={0.15} />
                )}
              </YStack>
            )
          })}
        </YStack>
      </Card>
    </YStack>
  )
}
