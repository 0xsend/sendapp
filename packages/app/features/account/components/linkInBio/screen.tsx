import {
  FadeCard,
  Paragraph,
  SubmitButton,
  Text,
  YStack,
  XStack,
  Input,
  Spinner,
  useMedia,
} from '@my/ui'
import { useCallback, useEffect, useState } from 'react'
import { FormProvider, useForm } from 'react-hook-form'
import { SettingsHeader } from 'app/features/account/components/SettingsHeader'
import { Platform } from 'react-native'
import {
  LinkInBioDomainNamesEnum,
  useLinkInBioMutation,
  type LinkInBioFormData,
} from 'app/utils/useLinkInBioMutation'
import { useUser } from 'app/utils/useUser'
import { IconLinkInBio } from 'app/components/icons'
import type { Database } from '@my/supabase/database-generated.types'

type LinkInBioDomainName = Database['public']['Enums']['link_in_bio_domain_names']

enum FormState {
  Overview = 'Overview',
  LinkInBioForm = 'LinkInBioForm',
}

export const LinkInBioScreen = () => {
  const media = useMedia()
  const { linksInBio, isLoading } = useUser()
  const form = useForm<LinkInBioFormData>()
  const { mutateAsync: updateLinkInBio } = useLinkInBioMutation()
  const [formState, setFormState] = useState<FormState>(FormState.Overview)
  const [errorMessage, setErrorMessage] = useState<string | null>(null)
  const [domainHandles, setDomainHandles] = useState<Record<LinkInBioDomainName, string>>(
    {} as Record<LinkInBioDomainName, string>
  )

  const handleSubmit = useCallback(async () => {
    setErrorMessage(null)

    try {
      // Convert domain handles to LinkInBioFormData format
      const linkInBioData: LinkInBioFormData = Object.entries(domainHandles || {}).map(
        ([domain_name, handle]) => ({
          domain_name: domain_name as LinkInBioDomainName,
          handle: handle.trim(),
        })
      )

      await updateLinkInBio(linkInBioData)
      setFormState(FormState.Overview)
    } catch (error) {
      console.error(error)
      if (error?.message) {
        setErrorMessage(error.message)
      }
    }
  }, [domainHandles, updateLinkInBio])

  useEffect(() => {
    if (linksInBio && Array.isArray(linksInBio)) {
      const handles: Record<LinkInBioDomainName, string> = {} as Record<LinkInBioDomainName, string>
      for (const link of linksInBio) {
        handles[link.domain_name] = link.handle ?? ''
      }
      setDomainHandles(handles)
    }
  }, [linksInBio])

  const updateDomainHandle = useCallback((domain_name: string, handle: string) => {
    setDomainHandles((prev) => ({
      ...prev,
      [domain_name]: handle,
    }))
  }, [])

  const renderAfterContent = useCallback(
    ({ submit }: { submit: () => void }) => (
      <YStack>
        <SubmitButton onPress={() => submit()}>
          <SubmitButton.Text>SAVE CHANGES</SubmitButton.Text>
        </SubmitButton>
        {errorMessage && (
          <Paragraph marginTop={'$5'} theme="red" color="$color9">
            {errorMessage}
          </Paragraph>
        )}
      </YStack>
    ),
    [errorMessage]
  )

  const socialLinksForm = (
    <YStack gap={'$5'}>
      <FadeCard elevation={'$0.75'}>
        {errorMessage && (
          <Paragraph marginBottom={'$5'} theme="red" color="$color9">
            {errorMessage}
          </Paragraph>
        )}
        <YStack gap={'$4'}>
          {Object.values(LinkInBioDomainNamesEnum).map((domain_name) => {
            const currentHandle = domainHandles[domain_name] || ''

            return (
              <YStack key={domain_name} gap={'$2'}>
                <XStack gap={'$4'} alignItems="center">
                  <IconLinkInBio domain_name={domain_name} />
                  <YStack flex={1} bbc={'$color12'}>
                    <Input
                      placeholder={domain_name}
                      placeholderTextColor={'$color10'}
                      value={currentHandle}
                      onChangeText={(value) => updateDomainHandle(domain_name, value)}
                      autoCapitalize="none"
                      autoCorrect={false}
                      bc={'$color0'}
                    />
                  </YStack>
                </XStack>
              </YStack>
            )
          })}
        </YStack>
      </FadeCard>
      {renderAfterContent({ submit: handleSubmit })}
    </YStack>
  )

  const overview = (
    <YStack gap={'$5'}>
      <FadeCard>
        {linksInBio && Array.isArray(linksInBio) && linksInBio.length > 0 ? (
          linksInBio.map(({ domain, handle, domain_name }, index) => {
            return (
              <YStack key={domain || `${domain_name}-${index}`}>
                <XStack gap={'$5'} ai="center">
                  <IconLinkInBio domain_name={domain_name} />
                  <Paragraph size={media.gtXs ? '$6' : '$4'} color={'$color12'}>
                    {`${domain}${handle ?? ''}`}
                  </Paragraph>
                </XStack>
              </YStack>
            )
          })
        ) : (
          <Text>No social links added yet.</Text>
        )}
      </FadeCard>
      <SubmitButton onPress={() => setFormState(FormState.LinkInBioForm)}>
        <SubmitButton.Text>edit links</SubmitButton.Text>
      </SubmitButton>
    </YStack>
  )

  if (isLoading) {
    return <Spinner size="small" color="$color12" />
  }

  return (
    <YStack w={'100%'}>
      <YStack gap={'$3.5'}>
        {Platform.OS === 'web' && <SettingsHeader>Link In Bio</SettingsHeader>}
        <FormProvider {...form}>
          {(() => {
            switch (formState) {
              case FormState.Overview:
                return overview
              case FormState.LinkInBioForm:
                return socialLinksForm
              default:
                return overview
            }
          })()}
        </FormProvider>
      </YStack>
    </YStack>
  )
}
