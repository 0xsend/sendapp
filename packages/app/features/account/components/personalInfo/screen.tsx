import {
  Button,
  FadeCard,
  isWeb,
  Paragraph,
  Separator,
  SubmitButton,
  Text,
  useToastController,
  YStack,
} from '@my/ui'
import { SchemaForm } from 'app/utils/SchemaForm'
import { useSupabase } from 'app/utils/supabase/useSupabase'
import { useUser } from 'app/utils/useUser'
import type { z } from 'zod'
import { FormProvider, useForm } from 'react-hook-form'
import { VerifyCode } from 'app/features/auth/components/VerifyCode'
import { AuthUserSchema, useAuthUserMutation } from 'app/utils/useAuthUserMutation'
import { useCallback, useEffect, useMemo, useState } from 'react'
import { useProfileMutation } from 'app/utils/useUserPersonalDataMutation'
import { useQuery } from '@tanstack/react-query'
import { adjustUTCDateForTimezone } from 'app/utils/dateHelper'
import { SettingsHeader } from 'app/features/account/components/SettingsHeader'
import { FieldWithLabel } from 'app/features/account/components/FieldWithLabel'
import { ReadOnlyFieldWithLabel } from 'app/features/account/components/ReadOnlyFieldWithLabel'

enum FormState {
  Overview = 'Overview',
  PersonalInfoForm = 'PersonalInfoForm',
  VerificationCode = 'VerificationCode',
}

export const PersonalInfoScreen = () => {
  const { profile, session } = useUser()
  const { data: user } = useQuery({
    queryKey: ['user'],
    enabled: !!session?.user,
    queryFn: async () => {
      const { data, error } = await supabase.auth.getUser()

      if (error) {
        await supabase.auth.signOut()
        throw new Error(error.message)
      }

      return data.user
    },
  })
  const supabase = useSupabase()
  const toast = useToastController()
  const form = useForm<z.infer<typeof AuthUserSchema>>() // Using react-hook-form
  const { mutateAsync: mutateAuthAsync } = useAuthUserMutation()
  const { mutateAsync: mutateProfileAsync } = useProfileMutation()
  const [formState, setFormState] = useState<FormState>(FormState.Overview)
  const [errorMessage, setErrorMessage] = useState<string | null>(null)

  async function handleSuccessCodeVerification() {
    setFormState(FormState.Overview)
    toast.show('Phone number updated')

    if (!isWeb) {
      await supabase.auth.refreshSession()
    }
  }

  async function handleUserUpdate() {
    const values = form.getValues()
    setFormState(FormState.VerificationCode)
    await mutateAuthAsync(values)
  }

  async function handleSubmit() {
    setErrorMessage(null)
    const values = form.getValues()
    const shouldUpdateUser = user?.phone !== values.phone

    try {
      await mutateProfileAsync(values)

      if (shouldUpdateUser) {
        await handleUserUpdate()
      } else {
        setFormState(FormState.Overview)
      }
    } catch (error) {
      console.error(error)

      if (error?.message) {
        setErrorMessage(error.message)
      }
    }
  }

  const birthday = useMemo(
    () => (profile?.birthday ? adjustUTCDateForTimezone(new Date(profile.birthday)) : undefined),
    [profile?.birthday]
  )

  const formatDate = (date?: Date) =>
    date?.toLocaleString(undefined, { day: 'numeric', month: 'long' }) || ''

  useEffect(() => {
    form.reset({
      phone: user?.phone ?? '',
      xUsername: profile?.x_username ?? '',
      birthday,
    })
  }, [profile?.x_username, user?.phone, form.reset, birthday])

  const verificationCode = (
    <FadeCard>
      <VerifyCode
        type={'phone_change'}
        phone={form.getValues().phone}
        onSuccess={handleSuccessCodeVerification}
      />
    </FadeCard>
  )

  const renderAfterContent = useCallback(
    ({ submit }: { submit: () => void }) => (
      <YStack>
        <SubmitButton theme="green" borderRadius={'$4'} p={'$4'} mt={'$1'} onPress={() => submit()}>
          <Button.Text ff={'$mono'} fontWeight={'500'} tt="uppercase" size={'$5'} color={'$black'}>
            SAVE CHANGES
          </Button.Text>
        </SubmitButton>
        {errorMessage && (
          <Paragraph marginTop={'$5'} theme="red" color="$color9">
            {errorMessage}
          </Paragraph>
        )}
      </YStack>
    ),
    [errorMessage]
  )

  const personalInfoForm = (
    <SchemaForm
      form={form}
      schema={AuthUserSchema}
      onSubmit={handleSubmit}
      props={{
        phone: {
          'aria-label': 'Phone number',
          autoComplete: 'tel',
          keyboardType: 'phone-pad',
          autoCapitalize: 'none',
          bc: '$color0',
        },
        xUsername: {
          'aria-label': 'X username',
          bc: '$color0',
          pl: '$8',
          iconBefore: (
            <Text color="$color10" userSelect={'none'} lineHeight={8}>
              @
            </Text>
          ),
        },
        birthday: {
          'aria-label': 'birthday',
          bc: '$color0',
          customDateFormatter: formatDate,
          disabled: Boolean(profile?.birthday),
        },
      }}
      formProps={{
        $gtSm: {
          maxWidth: '100%',
        },
      }}
      renderAfter={renderAfterContent}
    >
      {({ birthday, xUsername }) => (
        <FadeCard>
          <FieldWithLabel label={'Date of Birth'} additionalInfo={'(non-editable)'} gap={'$2'}>
            {birthday}
          </FieldWithLabel>
          <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
          <FieldWithLabel label={'X Handle'} gap={'$2'}>
            {xUsername}
          </FieldWithLabel>
        </FadeCard>
      )}
    </SchemaForm>
  )

  const overview = (
    <YStack gap={'$5'}>
      <FadeCard>
        <ReadOnlyFieldWithLabel
          label={'Date of Birth'}
          text={formatDate(birthday) || '-'}
          additionalInfo={'(non-editable)'}
        />
        <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
        <ReadOnlyFieldWithLabel
          label={'X Handle'}
          text={profile?.x_username ? `@ ${profile?.x_username}` : '-'}
        />
      </FadeCard>
      <SubmitButton
        theme="green"
        borderRadius={'$4'}
        p={'$4'}
        onPress={() => setFormState(FormState.PersonalInfoForm)}
      >
        <Button.Text ff={'$mono'} fontWeight={'500'} tt="uppercase" size={'$5'} color={'$black'}>
          edit personal information
        </Button.Text>
      </SubmitButton>
    </YStack>
  )

  return (
    <YStack w={'100%'}>
      <YStack gap={'$3.5'}>
        <SettingsHeader>Personal Information</SettingsHeader>
        <FormProvider {...form}>
          {(() => {
            switch (formState) {
              case FormState.Overview:
                return overview
              case FormState.PersonalInfoForm:
                return personalInfoForm
              case FormState.VerificationCode:
                return verificationCode
              default:
                return overview
            }
          })()}
        </FormProvider>
      </YStack>
    </YStack>
  )
}
