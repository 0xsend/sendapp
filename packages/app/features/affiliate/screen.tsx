import { Card, CardHeader, H3, H5, LinkableButton, Paragraph, XStack, YStack } from '@my/ui'
import { IconSendToken } from 'app/components/icons'
import { useUserReferralsCount } from 'app/utils/useUserReferralsCount'
import { api } from 'app/utils/api'
import { AnimateEnter } from '../home/TokenDetails'
import type { Tables } from '@my/supabase/database-generated.types'

export const AffiliateScreen = () => {
  return (
    <YStack f={1} width={'100%'} gap="$4">
      <XStack alignItems="center" width={'100%'} jc="flex-end" gap="$6">
        <LinkableButton href="/leaderboard" fontWeight={'bold'}>
          Check Leaderboard
        </LinkableButton>
      </XStack>
      <StatsCards />
      <History />
    </YStack>
  )
}

const StatsCards = () => {
  const { data: affiliateStats, error: affiliateStatsError } = useAffiliateStats()
  const { data: referralsCount, error: referralsCountError } = useUserReferralsCount()
  return (
    <XStack flexWrap="wrap" ai="flex-start" jc="space-around" gap="$8" mb="$4" width={'100%'}>
      <Card f={1} bc={'$background'}>
        <CardHeader>
          <Paragraph>Referrals</Paragraph>
        </CardHeader>
        <H5 pl="$4" pb="$3" fontWeight="600" size={'$7'}>
          {referralsCountError ? '?' : referralsCount || 0}
        </H5>
      </Card>
      <Card f={1} bc={'$background'}>
        <CardHeader>
          <Paragraph>Transactions</Paragraph>
        </CardHeader>
        <H5 pl="$4" pb="$3" fontWeight="600" size={'$7'}>
          {affiliateStatsError ? '?' : affiliateStats?.paymaster_tx_count || 0}
        </H5>
      </Card>
      <Card f={1} bc={'$background'}>
        <CardHeader>
          <Paragraph>Referral Transactions</Paragraph>
        </CardHeader>
        <H5 pl="$4" pb="$3" fontWeight="600" size={'$7'}>
          1000
        </H5>
      </Card>
    </XStack>
  )
}

const HistoryHeader = () => (
  <XStack gap="$1" ai="center">
    <Paragraph w="12%" mb="0" size="$5" lineHeight="$4">
      #
    </Paragraph>
    <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
      Sendtag
    </Paragraph>
    <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
      Referred Date
    </Paragraph>
  </XStack>
)

const ReferralsListRow = ({ referral }: { referral: Tables<'referrals'> }) => {
  return (
    <XStack width={'100%'} ai="center" jc="space-between" gap="$4" pb="$2">
      <XStack gap="$4.5" width={'100%'} f={1}>
      <Avatar size="$4.5" br="$4" gap="$2">
      <Avatar.Image
        src={referral.avatar_url ?? ''}
      />
      <Avatar.Fallback jc="center" bc="$olive">
        <Avatar size="$4.5" br="$4">
          <Avatar.Image
            src={'https://ui-avatars.com/api/?name=TODO&size=256&format=png&background=86ad7f'}
          />
        </Avatar>
      </Avatar.Fallback>
    </Avatar>
        <YStack gap="$1.5" width={'100%'} f={1} overflow="hidden">
          <XStack fd="row" jc="space-between" gap="$1.5" f={1} width={'100%'}>
            <Text color="$color12" fontSize="$6" $gtMd={{ fontSize: '$5' }}>
              {eventName}
            </Text>
            <Text color="$color12" fontSize="$5" ta="right">
              {amount}
            </Text>
          </XStack>
          <XStack
            gap="$1.5"
            alignItems="flex-start"
            justifyContent="space-between"
            width="100%"
            overflow="hidden"
            f={1}
          >
            {(isERC20Transfer || isETHReceive) &&
            Boolean(to_user?.send_id) &&
            Boolean(from_user?.send_id) ? (
              <Link
                href={`/profile/${
                  profile?.send_id === from_user?.send_id ? to_user?.send_id : from_user?.send_id
                }`}
              >
                <Paragraph
                  color="$color10"
                  fontFamily={'$mono'}
                  maxWidth={'100%'}
                  overflow={'hidden'}
                  fontSize="$4"
                  textDecorationLine="underline"
                >
                  {subtext}
                </Paragraph>
              </Link>
            ) : (
              <Paragraph
                color="$color10"
                fontFamily={'$mono'}
                maxWidth={'100%'}
                overflow={'hidden'}
                fontSize="$4"
              >
                {subtext}
              </Paragraph>
            )}
            <Paragraph color="$color10" size={'$3'}>
              {date}
            </Paragraph>
          </XStack>
        </YStack>
      </XStack>
    </XStack>
  )

const History = () => {
  return (
    <YStack f={1} width={'100%'} space="$4">
      <XStack alignItems="center" width={'100%'} gap="$3">
        <H3 fontWeight={'normal'}>Referrals</H3>
      </XStack>
      <YStack f={1} width={'100%'} p="$6" bg="$background" space="$4" br="$8">
        <HistoryHeader />
        <HistoryData />
      </YStack>
    </YStack>
  )
}

const HistoryData = () => {
  const data = ['data1', 'data2', 'data3']
  return data?.map((key) => (
    <XStack key={key} gap="$1" ai="center">
      <Paragraph w="12%" mb="0" size="$5" lineHeight="$4">
        ?
      </Paragraph>
      <Paragraph w="12%" f={1} mb="0" size="$5" lineHeight="$4" ta="center">
        ?
      </Paragraph>
      <XStack f={1} ai="center" w="12%" jc="center" mb="0" gap="$2">
        <IconSendToken />
        <Paragraph size="$5" lineHeight="$4" ta="center">
          ?
        </Paragraph>
      </XStack>
    </XStack>
  ))
}
