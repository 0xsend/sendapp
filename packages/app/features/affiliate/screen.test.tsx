import { expect, test } from '@jest/globals'
import { TamaguiProvider, config } from '@my/ui'
import { render, act, screen } from '@testing-library/react-native'
import { AffiliateScreen } from './screen'

jest.mock('app/utils/useUserReferralsCount', () => ({
  useUserReferralsCount: jest.fn().mockReturnValue({ data: 123, error: null }),
}))

test('AffiliateScreen', async () => {
  jest.useFakeTimers()
  render(
    <TamaguiProvider defaultTheme={'dark'} config={config}>
      <AffiliateScreen />
    </TamaguiProvider>
  )
  await act(async () => {
    jest.advanceTimersByTime(2000)
    jest.runAllTimers()
  })
  expect(screen.toJSON()).toMatchSnapshot()
})
