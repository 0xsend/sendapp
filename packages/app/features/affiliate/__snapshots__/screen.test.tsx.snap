// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AffiliateScreen 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "gap": 18,
      "paddingBottom": 32,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 32,
        "justifyContent": "flex-end",
        "width": "100%",
      }
    }
  >
    <View
      cancelable={true}
      disabled={false}
      focusVisibleStyle={{}}
      focusable={[Function]}
      minPressDuration={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onPress={[Function]}
      onPressIn={[Function]}
      onPressOut={[Function]}
      role="button"
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#081619",
          "borderBottomColor": "transparent",
          "borderBottomLeftRadius": 16,
          "borderBottomRightRadius": 16,
          "borderBottomWidth": 1,
          "borderLeftColor": "transparent",
          "borderLeftWidth": 1,
          "borderRightColor": "transparent",
          "borderRightWidth": 1,
          "borderStyle": "solid",
          "borderTopColor": "transparent",
          "borderTopLeftRadius": 16,
          "borderTopRightRadius": 16,
          "borderTopWidth": 1,
          "display": "flex",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "fontWeight": "bold",
          "height": 44,
          "justifyContent": "center",
          "paddingLeft": 18,
          "paddingRight": 18,
        }
      }
    >
      <Text
        lineBreakMode="clip"
        numberOfLines={1}
        style={
          {
            "color": "#FFFFFF",
            "flexGrow": 0,
            "flexShrink": 1,
            "fontFamily": "System",
            "fontSize": 16.8,
            "fontWeight": "bold",
            "lineHeight": 19,
            "userSelect": "none",
          }
        }
        suppressHighlighting={true}
      >
        Leaderboard
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flexDirection": "column",
        "flexWrap": "wrap",
        "gap": 13,
        "marginBottom": 18,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#111f22",
          "borderBottomLeftRadius": 16,
          "borderBottomRightRadius": 16,
          "borderTopLeftRadius": 16,
          "borderTopRightRadius": 16,
          "flexDirection": "column",
          "minHeight": 152,
          "position": "relative",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "transparent",
            "flexDirection": "column",
            "marginBottom": "auto",
            "paddingBottom": 18,
            "paddingLeft": 18,
            "paddingRight": 18,
            "paddingTop": 18,
            "zIndex": 10,
          }
        }
      >
        <Text
          cancelable={true}
          disabled={false}
          focusable={true}
          id=":r0:"
          minPressDuration={0}
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onPress={[Function]}
          onPressIn={[Function]}
          onPressOut={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "color": "#B2B2B2",
              "display": "flex",
              "fontFamily": "System",
              "fontSize": 16.8,
              "fontWeight": "400",
              "lineHeight": 44,
              "userSelect": "none",
            }
          }
          suppressHighlighting={true}
        >
          Total Referrals
        </Text>
      </View>
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontFamily": "System",
            "fontSize": 74.39999999999999,
            "fontWeight": "600",
            "lineHeight": 60,
            "paddingBottom": 13,
            "paddingLeft": 18,
            "userSelect": "auto",
          }
        }
        suppressHighlighting={true}
      >
        0
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 13,
        }
      }
    >
      <Text
        accessibilityRole="header"
        style={
          {
            "color": "#FFFFFF",
            "fontFamily": "System",
            "fontSize": 23,
            "fontWeight": "normal",
            "lineHeight": 27,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textTransform": "none",
            "userSelect": "auto",
          }
        }
        suppressHighlighting={true}
      >
        No Referrals
      </Text>
    </View>
  </View>
</View>
`;
